README.rst
PySIS
=============================

What is it?
-----------------------------------------
Student Information System, developed with Django and Python.
This software is still in development. Not yet ready for production use.


Latest Version
-----------------------------------------
The latest version of this project can be found at : http://bitbucket.org/sramana/pysis.

Currently deployed at http://sis.rmv.ac.in


Documentation
-----------------------------------------
The documentation is distributed along with source repository in docs/ directory (in ReStructuredText format). You can run "make html" in that directory to generate HTML version.

Alternatively, you can browse the online documentation at http://sramana.github.com/pysis


Installation
-----------------------------------------
Installation instructions are provided in docs/install.rst


Contribution and Feedback
-----------------------------------------
Contributions and Feedback are most welcome. Please email the author with your comments.


Author Information
-----------------------------------------
Ramana <sramana9@gmail.com>

* http://bitbucket.org/sramana
* http://github.com/sramana


(Un)license
-----------------------------------------
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>


CHANGELOG
Revision 135 	 2011-10-01 	 Fixing css not loading issue 
Revision 134 	 2011-09-30 	 Using help.srkv.org 
Revision 133 	 2011-08-21 	 Fixing uwsgi timeout issue for long running requests 
Revision 132 	 2011-08-20 	 After renaming, Google Apps is putting original username as nickname. We don't want it 
Revision 131 	 2011-08-20 	 Bumping nginx proxy_read_timeout 
Revision 130 	 2011-08-20 	 Fix for error in pre_save 
Revision 129 	 2011-08-20 	 Fixing import pre_save 
Revision 128 	 2011-08-20 	 Making college_email_id unique 
Revision 127 	 2011-08-20 	 Fix for Django's bug regarding not being able to have more than one row with NULL in CharField with unique=True, blank=True, null=True 
Revision 126 	 2011-08-20 	 Adding college_email_id to search fileds 
Revision 125 	 2011-08-20 	 Bumping the number of uwsgi processes to 2 
Revision 124 	 2011-08-20 	 changing passwd in new accounts to use digits only. Easy for students to notedown the new password 
Revision 123 	 2011-08-20 	 Adding a unique digit after username if unique username is not found 
Revision 122 	 2011-08-20 	 changing reset_password to use digits only. Easy for students to notedown the new password 
Revision 121 	 2011-08-11 	 Never delete an account from Google Apps. Just rename it. 
Revision 120 	 2011-05-05 	 Enabling Django Compressor 
Revision 119 	 2011-05-04 	 Cleaning up fabfile 
Revision 118 	 2011-05-04 	 Setting EMAIL_BACKEND to locmem in test_settings 
Revision 117 	 2011-05-04 	 Updating fabfile to make nosetests pass 
Revision 116 	 2011-05-04 	 Fixing favicon url 
Revision 115 	 2011-05-04 	 Increasing uWSGI buffer size to 32768 to fix 'invalid request block size' error 
Revision 114 	 2011-05-04 	 Adding code to serve static files in dev 
Revision 113 	 2011-05-04 	 Fixing requirements file 
Revision 112 	 2011-05-04 	 Removing pysis_facgi.py 
Revision 111 	 2011-05-04 	 Major Rehaul : Using Nginx + uWSGI instead of Nginx + flup + FastCGI for deployment 
Revision 110 	 2011-05-04 	 Upgrading settings to Django 1.3 version. Primary diff : static and media 
Revision 109 	 2011-05-04 	 Renaming MEDIA_URL to STATIC_URL in templates as per Django 1.3 nomenclature 
Revision 108 	 2011-05-04 	 Adding manage.py 
Revision 107 	 2011-05-04 	 Renaming media to static as per Django 1.3 nomenclature 
Revision 106 	 2011-05-04 	 Updating requirements 
Revision 105 	 2011-03-04 	 Adding apache config 
Revision 104 	 2011-03-02 	 Adding email id to birthday reminders 
Revision 103 	 2011-03-02 	 Adding email id to birthday reminders 
Revision 102 	 2010-12-25 	 handling duplicate entries whil bulk loading 
Revision 101 	 2010-12-25 	 Fix for null last_name 
Revision 100 	 2010-12-25 	 Fix for null last_name 
Revision 99 	 2010-12-25 	 Removing link to Apps 
Revision 98 	 2010-12-25 	 migrating from IdeaScale to FormSpring 
Revision 97 	 2010-12-25 	 Config related changes 
Revision 96 	 2010-11-26 	 Name change 
Revision 95 	 2010-11-23 	 Fix for groupname of staff mebers 
Revision 94 	 2010-11-23 	 Fix for groupname of staff mebers 
Revision 93 	 2010-11-21 	 Minor changes 
Revision 92 	 2010-11-18 	 Changing author name 
Revision 91 	 2010-11-13 	 Adding IE6NoMore 
Revision 90 	 2010-11-06 	 Doc related changes 
Revision 89 	 2010-11-02 	 Correction to install instructions 
Revision 88 	 2010-11-02 	 Small fix for git push in docs 
Revision 87 	 2010-11-01 	 Documentation related changes 
Revision 86 	 2010-11-01 	 Chaning hashlib import to make pylint happy 
Revision 85 	 2010-10-27 	 Small correction 
Revision 84 	 2010-10-27 	 removing version number of gdata from requirements as it is causing errors 
Revision 83 	 2010-10-27 	 Deployment related cleanup 
Revision 82 	 2010-10-21 	 Upgrading latest versions of requirements 
Revision 81 	 2010-10-21 	 Changes to requirements 
Revision 80 	 2010-10-19 	 Initialized Ditz bug database 
Revision 79 	 2010-10-19 	 Refactored some config stuff 
Revision 78 	 2010-09-15 	 Prettifying Student profile listing pages 
Revision 77 	 2010-09-09 	 Warning when navigating away from unsaved edit form 
Revision 76 	 2010-09-05 	 Cleanup of fabfile.py 
Revision 75 	 2010-09-05 	 Changing homepage for logged-in users 
Revision 74 	 2010-09-05 	 Changing theme of docs to natures as readthedocs.org does not support custom themes 
Revision 73 	 2010-09-04 	 Small change to docs 
Revision 72 	 2010-09-04 	 Using ADCtheme for documentation 
Revision 71 	 2010-08-30 	 Renaming README to README.rst 
Revision 70 	 2010-08-24 	 Chaging URL in main menu 
Revision 69 	 2010-08-24 	 Logout now logs out of Google Apps also 
Revision 68 	 2010-08-24 	 Small change to main menu 
Revision 67 	 2010-08-23 	 Password reset confirm form 
Revision 66 	 2010-08-23 	 Removing forcePasswordChange as it is not working 
Revision 65 	 2010-08-23 	 Removing forcePasswordChange as it is not working 
Revision 64 	 2010-08-23 	 small testing related changes 
Revision 63 	 2010-08-23 	 Small changes to password reset 
Revision 62 	 2010-08-22 	 Small change to testing 
Revision 61 	 2010-08-22 	 Added testcases for password reset and greetings 
Revision 60 	 2010-08-22 	 Greetings on Birthdays. Not tested 
Revision 59 	 2010-08-22 	 Added Password Reset. Not tested 
Revision 58 	 2010-08-22 	 Testing related changes 
Revision 57 	 2010-08-17 	 Google Apps related changes. Tested 
Revision 56 	 2010-08-16 	 Small cosmetic changes 
Revision 55 	 2010-08-16 	 Customizing Admin Dashboard 
Revision 54 	 2010-08-16 	 To Activate/Deactivate accounts in Google 
Revision 53 	 2010-08-16 	 Confirmation while deleting Google account 
Revision 52 	 2010-08-16 	 Small changes to admin 
Revision 51 	 2010-08-16 	 Changing vidyalaya_email_id to college_email_id 
Revision 50 	 2010-08-16 	 sending email on activation 
Revision 49 	 2010-08-16 	 Google Apps realted changes. Not tested 
Revision 48 	 2010-08-14 	 Adding related_name to profile user 
Revision 47 	 2010-08-14 	 Mistake in settings 

apps/accounts/admin.py
from django.contrib import admin
from django.contrib import databrowse
from django.contrib import messages
from django.conf import settings
from django.contrib.admin.util import get_deleted_objects
from django.shortcuts import render_to_response
from django.utils.encoding import force_unicode
from django.contrib.admin import helpers
from django import template

import fullhistory
from fullhistory.admin import FullHistoryAdmin

from django.contrib.auth.models import User
from accounts.models import Course, Batch, Profile
from accounts.google_apps_manager import GoogleAppsManager
from accounts import accounts_manager

def register(model, modelAdmin):
    admin.site.register(model, modelAdmin)
    databrowse.site.register(model)
    fullhistory.register_model(model)

class DefaultAdmin(FullHistoryAdmin):
    pass

class ProfileAdmin(FullHistoryAdmin):

    # Options for admin
    list_display = ('full_name',
                    'user',
                    'college_email_id',
                    'personal_email_id',
                    'personal_contact_number',
                    'actual_date_of_birth',
                    'google_account_created',
                    'active',
                    'last_modified_on',)
    #list_editable = ('college_email_id',)

    list_filter = ('google_account_created',
                   'active',
                   'year_of_joining',
                   'course',
                   'blood_group',
                   'reservation_category',
                   )
    list_per_page = 20
    search_fields = ('user__first_name',
                     'user__last_name',
                     'user__username',
                     'personal_email_id',
                     'college_email_id')
    actions = ['create_accounts_in_google',
               'delete_accounts_from_google',
               'populate_college_email_id',
               'reset_password',
               'mark_as_processed',
               'mark_as_not_processed',
               'deactivate_account',
               'reactivate_account',
              ]

    fieldsets = ((None,
                  {'fields': ('user', 'course', 'year_of_joining', 'college_email_id',)
                  }
                 ),
                 ('Personal Details',
                  {'classes': ('collapse','closed',),
                   'fields': settings.PERSONAL_DETAILS_FIELD_LIST
                  }
                 ),
                 ('Family Details',
                  {'classes': ('collapse','closed',),
                   'fields': settings.FAMILY_DETAILS_FIELD_LIST
                  }
                 ),
                 ('Contact Details',
                  {'classes': ('collapse','closed',),
                   'fields': settings.CONTACT_DETAILS_FIELD_LIST
                  }
                 ),
                 ('Education Details',
                  {'classes': ('collapse','closed',),
                   'fields': settings.EDUCATION_DETAILS_FIELD_LIST
                  }
                 ),
                 ('Misc Details',
                  {'classes': ('collapse','closed',),
                   'fields': settings.MISC_DETAILS_FIELD_LIST
                  }
                 ),
                )


    def create_accounts_in_google(self, request, queryset):
        """Creates the Google Apps account for the user
        """
        gam = GoogleAppsManager()
        password = User.objects.make_random_password(length=8, allowed_chars='0123456789')

        for profile in queryset:
            accounts_manager.create_account_in_google_apps(request, profile, password)


    def delete_accounts_from_google(self, request, queryset):
        """Deletes the user from Google Apps database
        """
        gam = GoogleAppsManager()

        opts = self.model._meta
        app_label = opts.app_label

        deletable_objects = [profile for profile in queryset]

        # The user has already confirmed the deletion.
        # Do the deletion and return a None to display the change list view again.
        if request.POST.get('post'):
            for profile in queryset:
                try:
                    gam.delete_account(profile.user.username)
                    profile.google_account_created = False
                    profile.save()
                except Exception, e:
                    messages.error(request,
                        'Error while deleting %s. Error : %s' %
                        (profile.register_number, e))
                else:
                    messages.success(request,
                        'Successfully deleted %s' % profile.register_number)
            return None

        context = {
            "title": "Are you sure?",
            "object_name": force_unicode(opts.verbose_name),
            "deletable_objects": [deletable_objects],
            'queryset': queryset,
            "perms_lacking": False,
            "opts": opts,
            "root_path": self.admin_site.root_path,
            "app_label": app_label,
            'action_checkbox_name': helpers.ACTION_CHECKBOX_NAME,
        }

        # Display the confirmation page
        return render_to_response('accounts/delete_from_google_apps_confirmation.html',
                                  context,
                                  context_instance=template.RequestContext(request))


    def populate_college_email_id(self, request, queryset):
        """Computes unique email id and populates
        """
        for profile in queryset:
            # Populate only if it is empty.
            if profile.college_email_id:
                messages.error(request,
                    'College email id is already populated for %s. Not modifying.' % profile.register_number)
            else:
                username = accounts_manager.get_new_username(profile.user.first_name,
                                                             profile.user.last_name)

                if username:
                    profile.college_email_id = username + '@' + settings.GOOGLE_APPS_DOMAIN
                    profile.save()
                else:
                    messages.error(request,
                        'Could not generate a unique username for %s' % profile.register_number)


    def reset_password(self, request, queryset):
        gam = GoogleAppsManager()
        passwd = User.objects.make_random_password(length=8, allowed_chars='0123456789')

        for profile in queryset:
            if not profile.google_account_created:
                messages.error(request,
                    'No Google Apps account for %s' % profile.register_number)
                continue

            try:
                username = profile.register_number
                result = gam.change_password(username,
                                             passwd)
            except Exception, e:
                messages.error(request,
                    'Failed to update password for %s. Reason : %s' %
                    (username, e))
            else:
                messages.success(request,
                    'Successfully updated password for %s. New Password is %s' %
                    (username, passwd))

    def mark_as_processed(self, request, queryset):
        queryset.update(google_account_created=True)

    def mark_as_not_processed(self, request, queryset):
        queryset.update(google_account_created=False)

    def deactivate_account(self, request, queryset):
        gam = GoogleAppsManager()

        for profile in queryset:
            try:
                gam.suspend_user(profile.user.username)
                profile.active = False
                profile.save()
            except Exception, e:
                messages.error(request,
                    'Error while deactivating %s. Reason : %s' %
                    (profile.user.username, e))
            else:
                messages.success(request,
                    'Deactivated %s' % profile.user.username)

    def reactivate_account(self, request, queryset):
        gam = GoogleAppsManager()

        for profile in queryset:
            try:
                gam.unsuspend_user(profile.user.username)
                profile.active = True
                profile.save()
            except Exception, e:
                messages.error(request,
                    'Error while reactivating %s. Reason : %s' %
                    (profile.user.username, e))
            else:
                messages.success(request,
                    'Reactivated %s' % profile.user.username)


register(Course, DefaultAdmin)
register(Batch, DefaultAdmin)
register(Profile, ProfileAdmin)


settings/pysis_settings.py
# These are settings specific to PySIS project

import pysis
REVISION = pysis.revision

# We need this to match our css classes
from django.contrib import messages
MESSAGE_TAGS = {
    messages.DEBUG : 'info',
    messages.INFO : 'info',
    messages.SUCCESS : 'done',
    messages.WARNING : 'warning',
    messages.ERROR : 'error',
}

AUTH_PROFILE_MODULE = 'accounts.Profile'

# Google Apps Settings
GOOGLE_APPS_DOMAIN = 'rmv.ac.in'
GOOGLE_APPS_ADMIN_USERNAME = 'admin@mydomain.com'
GOOGLE_APPS_ADMIN_PASSWORD = 'my_secret_pass'

GOOGLE_ANALYTICS_ID = "'UA-15662130-6'"

MY_PROFILE_LANDING_URL = '/myprofile/general/'
STUDENTS_LANDING_URL = '/students/browse/myclassmates/'

LOGIN_URL = '/openid/login/'
LOGIN_REDIRECT_URL = MY_PROFILE_LANDING_URL
LOGOUT_REDIRECT_URL = 'https://www.google.com/accounts/Logout?' +\
                      'continue=https://www.google.com/a/' +\
                      GOOGLE_APPS_DOMAIN
PASSWORD_CHANGE_URL = 'https://www.google.com/accounts/EditPasswd'
HOMEPAGE_FOR_LOGGEDIN_USERS = MY_PROFILE_LANDING_URL

MAIN_MENU_LEFT = [

    # (Name, URL)
    ('My Profile', '/myprofile/'),
    ('Student Profiles', '/students/'),
    ('Attendance', '/attendance/'),
    ('Marks', '/marks/'),
    ('Library', '/library/'),
]

MAIN_MENU_RIGHT = [

    # (Name, URL)
    ('Help', 'http://help.srkv.org'),
]

MY_PROFILE_MENU = [

    # (Name, URL)
    ('General Details', '/myprofile/general/'),
    ('Personal Details', '/myprofile/personal/'),
    ('Family Details', '/myprofile/family/'),
    ('Contact Details', '/myprofile/contact/'),
    ('Education Details', '/myprofile/education/'),
    ('Misc Details', '/myprofile/misc/'),
]


STUDENTS_PROFILE_MENU = [

    # (Name, URL)
    ('My Classmates', '/students/browse/myclassmates/'),
    ('My Seniors', '/students/browse/myseniors/'),
    ('My Juniors', '/students/browse/myjuniors/'),
    ('', ''),
    ('Browse by Batch', '/students/browse/'),
    ('', ''),
    ('All Students', '/students/search/?q=all'),
]

STUDENTS_PROFILE_MENU_URLS =  zip(*STUDENTS_PROFILE_MENU)[1]

# These fields determine who can see what.
# For the complete list of available fields, see myprofile.views.Profile

PROFILE_PREVIEW_FIELD_LIST = [
    'register_number',
    'course',
    'birth_day',
    'personal_email_id',
    'college_email_id',
    'personal_contact_number',
    ]

PROFILE_FULLVIEW_FIELD_LIST = [
    'full_name',
    'register_number',
    'course',
    'birth_day',
    'personal_email_id',
    'personal_email_id2',
    'college_email_id',
    'personal_contact_number',
    'personal_website',
    'orkut_profile_url',
    'facebook_profile_url',
    'linkedin_profile_url',
    ]

GENERAL_DETAILS_FIELD_LIST = [
    'first_name',
    'last_name',
    'register_number',
    'course',
    'year_of_joining',
    'college_email_id',
    ]

PERSONAL_DETAILS_FIELD_LIST = [
    'gender',
    'blood_group',
    'date_of_birth',
    'actual_date_of_birth',
    ]

FAMILY_DETAILS_FIELD_LIST = [
    'father_name',
    'father_profession',
    'family_income',
    'religion',
    'caste',
    'reservation_category',
    ]

CONTACT_DETAILS_FIELD_LIST = [
    'personal_email_id',
    'personal_email_id2',
    'personal_contact_number',
    'personal_contact_number2',
    'emergency_contact_number',
    'present_address',
    'permanent_address',
    ]

EDUCATION_DETAILS_FIELD_LIST = [
    'sslc_passed_out_year',
    'sslc_percentage',
    'hslc_passed_out_year',
    'hslc_percentage',
    'hslc_major',
    'ug_passed_out_year',
    'ug_percentage',
    'ug_major',
    'pg_passed_out_year',
    'pg_percentage',
    'pg_major',
    'history_of_arrears',
    ]

MISC_DETAILS_FIELD_LIST = [
    'nss',
    'ncc',
    'sports',
    'hobbies',
    'personal_website',
    'orkut_profile_url',
    'facebook_profile_url',
    'linkedin_profile_url',
    ]


apps/accounts/accounts_manager.py
#!/usr/bin/env python

from pysis.scripts import bootstrap_django
bootstrap_django.main()

import csv
from django.db.utils import IntegrityError
from django.db import IntegrityError
from django.conf import settings
from django.contrib import messages
from django.contrib.auth.models import User

from accounts.models import Profile, Course
from accounts.google_apps_manager import GoogleAppsManager
from generic_app.email import send_html_mail

def get_names(full_name):
    """ Sanitizes the full_name and
        returns first_name and last_name
    """
    full_name = full_name.replace('.', ' ') # Remove Dot
    parts = full_name.split() # Remove extra whitespace and split into words

    first_name = ''
    last_name = ''
    for part in parts:
        if len(part) == 1: # Assuming that last_name has only one character
            last_name = last_name + ' ' + part
        else:
            first_name = first_name + ' ' + part

    if not last_name:
        # if there is no last name, use last part of first name
        last_name = first_name.split()[-1]
        first_name = first_name.replace(last_name, '').strip()

    if not first_name:
        # if there is no last name, use last part of first name
        first_name = last_name

    first_name = first_name.strip().title()
    last_name = last_name.strip().title()

    return first_name, last_name

def is_unique(username):
    domain = settings.GOOGLE_APPS_DOMAIN
    email_id = username + '@' + domain

    exists = Profile.objects.filter(college_email_id__iexact=email_id).exists()

    return not exists

def get_new_username(first_name,
                     last_name):
    """Returns a new, unique username for Google Apps based on the following rules:
        email_id = first word of first_name
        email_id = complete first_name without spaces
        email_id = first word of first_name + dot + first letter of last_name
        email_id = first word of first_name + dot + first word of last_name
        email_id = complete first_name without spaces + dot + first letter of last_name
        email_id = complete first_name without spaces + dot + complete last_name without spaces
        email_id = blank

    If can not find a unique email id by following these rules, return null.
    Admin has to manually set it.
    """

    first_name = first_name.lower()
    last_name = last_name.lower()
    first_word_of_first_name = first_name.split()[0]
    first_word_of_last_name = last_name.split()[0]
    first_letter_of_last_name = last_name[0]
    first_name_without_spaces = ''.join(first_name.split())
    last_name_without_spaces = ''.join(last_name.split())

    username = first_word_of_first_name
    if  is_unique(username):
        return username

    username = first_name_without_spaces
    if is_unique(username):
        return username

    username = first_word_of_first_name + '.' + first_letter_of_last_name
    if is_unique(username):
        return username

    username = first_word_of_first_name + '.' + first_word_of_last_name
    if is_unique(username):
        return username

    username = first_name_without_spaces + '.' + first_letter_of_last_name
    if is_unique(username):
        return username

    username = first_name_without_spaces + '.' + last_name_without_spaces
    if is_unique(username):
        return username

    for i in range(1,100):
        username = first_word_of_first_name + str(i)
        if is_unique(username):
            return username


def send_introduction_mail(first_name,
                           college_email_id):

    c = {'first_name': first_name,}
    subject = "Welcome to %s" % settings.ORGANIZATION

    send_html_mail('accounts/introduction_email.html',
                   c, subject, college_email_id)


def create_account_in_google_apps(request, profile, password):

    username = profile.user.username
    first_name = profile.user.first_name
    last_name = profile.user.last_name
    college_email_id = profile.college_email_id
    nickname = college_email_id.split('@')[0]

    if profile.course.name == 'Staff':
        groupname = 'staff'
    else:
        groupname = profile.register_number[:5]

    if not college_email_id:
        messages.error(request,
             'College Email Id is empty for %s' % username)
        return

    if profile.google_account_created:
        messages.error(request,
             'College Email Id is already created for %s' % username)
        return

    try:
        gam = GoogleAppsManager()

        gam.create_account(username,
                           password,
                           first_name,
                           last_name,
                           nickname,
                           groupname
                          )

        send_introduction_mail(first_name, college_email_id)

        profile.google_account_created = True
        profile.save()

    except Exception, e:
        messages.error(request,
            'Error while creating %s. Error : %s' %
            (profile.register_number, e))
    else:
        messages.success(request,
            'Successfully created %s. Password is %s' %
            (profile.register_number, password))


def create_accounts(students):
    for register_number, name in students:
        username = register_number.lower()
        password = None
        first_name, last_name = get_names(name)
        domain = settings.GOOGLE_APPS_DOMAIN
        email = username + '@' + domain

        # First two letters of register_number represents year
        year_of_joining = register_number[:2].replace('m', '0')
        year_of_joining = int(year_of_joining) + 2000

        course_code =  register_number[2:5]
        course = Course.objects.get(code__iexact=course_code)

        try:
            user = User.objects.create_user(username=username,
                                            password=password,
                                            email=email)

            user.first_name = first_name
            user.last_name = last_name
            user.save()

        except IntegrityError:
            print 'Account %s already exists' % username
            user = User.objects.get(username=username)

        try:
            profile = Profile.objects.create(user=user,
                                             course=course,
                                             year_of_joining=year_of_joining)
        except IntegrityError:
            print 'Profile %s already exists' % username

        print 'Successfully Created %s' % username


if __name__ == '__main__':
    # Output of
    # select distinct reg_no, name from Result where reg_no like '08MCA%' order by reg_no;
    CSV_FILE = '/tmp/students.csv'
    students = csv.reader(open(CSV_FILE), delimiter='|', skipinitialspace=True)
    create_accounts(students)


apps/accounts/models.py
import datetime, calendar

from django.db import models
from django.conf import settings
from django.contrib.auth.models import User

from managers import ProfileManager


# Fix for Django's bug regarding not being able to have more than one row with NULL in CharField with unique=True, blank=True, null=True
# Taken from https://www.maniacmartin.com//2010/12/21/unique-nullable-charfields-django/
from django.db.models.signals import pre_save
def cleanup(sender, instance, **kwargs):
    for mf in instance._meta.fields:
        if mf.null and mf.blank:
            f = instance.__getattribute__(mf.name)
            if f == "":
                f = None
pre_save.connect(cleanup)


GENDER_CHOICES = (
    ('M', 'Male'),
    ('F', 'Female'),
    )

RESERVATION_CATEGORY_CHOICES = (
    ('OC', 'OC'),
    ('OBC', 'OBC'),
    ('MBC', 'MBC'),
    ('BC', 'BC'),
    ('SC', 'SC'),
    ('ST', 'ST'),
    )

BLOOD_GROUP_CHOICES = (
    ('O+', 'O+'),
    ('O-', 'O-'),
    ('A+', 'A+'),
    ('A-', 'A-'),
    ('B+', 'B+'),
    ('B-', 'B-'),
    ('AB+', 'AB+'),
    ('AB-', 'AB-'),
    )

COURSE_LEVELS = (
    ('UG', 'Under Graduate'),
    ('PG', 'Post Graduate'),
    )

class Course(models.Model):
    code = models.CharField(max_length=10, unique=True)
    name = models.CharField(max_length=50, unique=True)
    duration = models.IntegerField(help_text='Duration of the course in years')
    level = models.CharField(max_length=4,
                             choices=COURSE_LEVELS,
                             default='UG')

    objects = models.Manager()

    def __unicode__(self):
        return self.name

class Batch(models.Model):
    code = models.CharField(max_length=5, unique=True,
                            help_text='First 5 characters of register number')
    course = models.ForeignKey(Course, null=True, blank=True)
    year = models.IntegerField()
    name = models.CharField(max_length=50, unique=True)

    def __unicode__(self):
        return self.name

    class Meta:
        ordering = ["course", "-year"]
        verbose_name_plural = "Batches"


class Profile(models.Model):
    # Register Number, First Name and Last Name are defined in User model
    user = models.ForeignKey(User, unique=True,)

    objects = ProfileManager()

    @property
    def first_name(self):
        return self.user.first_name

    @property
    def last_name(self):
        return self.user.last_name

    @property
    def full_name(self):
        return '%s %s' % (self.user.first_name, self.user.last_name)

    @property
    def register_number(self):
        return self.user.username

    @property
    def last_login(self):
        return self.user.last_login

    @property
    def birth_day(self):
        if self.actual_date_of_birth:
            day = self.actual_date_of_birth.day
            month = calendar.month_name[self.actual_date_of_birth.month]
            return '%s %s' % (month, day)

    # GeneralDetails
    course = models.ForeignKey(Course, null=True, blank=True)
    year_of_joining = models.DecimalField(max_digits=4, decimal_places=0, null=True, blank=True)
    college_email_id = models.EmailField(max_length=75, null=True, blank=True, unique=True)


    # PersonalDetails
    gender = models.CharField(max_length=2,
                              choices=GENDER_CHOICES,
                              default='M')
    blood_group = models.CharField(max_length=4, choices=BLOOD_GROUP_CHOICES,null=True, blank=True)
    date_of_birth = models.DateField(help_text='What is your date of birth as per records?',null=True, blank=True)
    actual_date_of_birth = models.DateField(help_text='When do you celebrate your birthday?',null=True, blank=True)


    # FamilyDetails
    father_name = models.CharField(max_length=100, null=True, blank=True)
    father_profession = models.CharField(max_length=100, null=True, blank=True)
    family_income = models.IntegerField(null=True, blank=True)
    religion = models.CharField(max_length=30,
                             null=True, blank=True)
    caste = models.CharField(max_length=30,
                             null=True, blank=True)
    reservation_category = models.CharField(max_length=10,
                                            choices=RESERVATION_CATEGORY_CHOICES,
                                            null=True, blank=True)


    # Contact Details
    personal_email_id = models.EmailField(max_length=75, null=True, blank=True)
    personal_email_id2 = models.EmailField(max_length=75, null=True, blank=True)
    personal_contact_number = models.CharField(max_length=20, null=True, blank=True)
    personal_contact_number2 = models.CharField(max_length=20, null=True, blank=True)
    emergency_contact_number = models.CharField(max_length=20, null=True, blank=True)
    present_address = models.TextField(null=True, blank=True)
    permanent_address = models.TextField(null=True, blank=True)


    # EducationDetails
    sslc_passed_out_year = models.DecimalField(max_digits=4, decimal_places=0, null=True, blank=True)
    sslc_percentage = models.DecimalField(max_digits=4, decimal_places=2, null=True, blank=True)

    hslc_passed_out_year = models.DecimalField(max_digits=4, decimal_places=0, null=True, blank=True)
    hslc_percentage = models.DecimalField(max_digits=4, decimal_places=2, null=True, blank=True)
    hslc_major = models.CharField(max_length=30, null=True, blank=True)

    ug_passed_out_year = models.DecimalField(max_digits=4, decimal_places=0, null=True, blank=True)
    ug_percentage = models.DecimalField(max_digits=4, decimal_places=2, null=True, blank=True,
                                        help_text='Percentage till now')
    ug_major = models.CharField(max_length=30, null=True, blank=True)

    pg_passed_out_year = models.DecimalField(max_digits=4, decimal_places=0, null=True, blank=True)
    pg_percentage = models.DecimalField(max_digits=4, decimal_places=2, null=True, blank=True,
                                        help_text='Percentage till now')
    pg_major = models.CharField(max_length=30, null=True, blank=True)

    history_of_arrears = models.TextField(null=True, blank=True)


    # Misc Details
    nss = models.BooleanField(blank=True)
    ncc = models.BooleanField(blank=True)
    sports = models.CharField(max_length=100, null=True, blank=True)
    hobbies = models.CharField(max_length=100, null=True, blank=True)
    personal_website = models.URLField(null=True, blank=True, verify_exists=False)
    orkut_profile_url = models.URLField(null=True, blank=True, verify_exists=False)
    facebook_profile_url = models.URLField(null=True, blank=True, verify_exists=False)
    linkedin_profile_url = models.URLField(null=True, blank=True, verify_exists=False)

    # Audit Trail
    active = models.BooleanField(default=True)
    google_account_created = models.BooleanField(default=False)
    last_modified_on = models.DateTimeField(editable=False,
                                            auto_now=True)

    def __unicode__(self):
        return '%s %s - %s' % (self.user.first_name, self.user.last_name, self.user.username )
        #return self.user.username

    def slice(self, fields_list):
        slice = []

        for field in fields_list:
            slice.append((field.replace('_', ' '), self.__getattribute__(field)))

        return slice

    def general_details(self):
        return self.slice(settings.GENERAL_DETAILS_FIELD_LIST)

    def personal_details(self):
        return self.slice(settings.PERSONAL_DETAILS_FIELD_LIST)

    def family_details(self):
        return self.slice(settings.FAMILY_DETAILS_FIELD_LIST)

    def contact_details(self):
        return self.slice(settings.CONTACT_DETAILS_FIELD_LIST)

    def education_details(self):
        return self.slice(settings.EDUCATION_DETAILS_FIELD_LIST)

    def misc_details(self):
        return self.slice(settings.MISC_DETAILS_FIELD_LIST)

    def search_preview(self):
        return self.slice(settings.PROFILE_PREVIEW_FIELD_LIST)

    def search_fullview(self):
        return self.slice(settings.PROFILE_FULLVIEW_FIELD_LIST)

    class Meta:
        ordering = ['-last_modified_on']
